rtengine/imageio.cc

```diff
@@ -822,6 +829,171 @@ int ImageIO::loadTIFF (const Glib::ustring &fname)
+                runner.get(),
+                JxlResizableParallelRunnerSuggestThreads(info.xsize, info.ysize));
+        } else if (status == JXL_DEC_COLOR_ENCODING) {
+            // check for ICC profile
+            deleteLoadedProfileData();
+            embProfile = nullptr;
+            size_t icc_size = 0;
+
+            if (JXL_DEC_SUCCESS !=
+#if JPEGXL_NUMERIC_VERSION < JPEGXL_COMPUTE_NUMERIC_VERSION(0,8,0)
+                JxlDecoderGetICCProfileSize(dec.get(), &format, _PROFILE_, &icc_size)
+#else
+                JxlDecoderGetICCProfileSize(dec.get(), _PROFILE_, &icc_size)
+#endif
+            ) {
+                g_printerr("Warning: JxlDecoderGetICCProfileSize failed\n");
+            }
+
+            if (icc_size > 0) {
+                icc_profile.resize(icc_size);
+                if (JXL_DEC_SUCCESS !=
+#if JPEGXL_NUMERIC_VERSION < JPEGXL_COMPUTE_NUMERIC_VERSION(0,8,0)
+                    JxlDecoderGetColorAsICCProfile(
+                        dec.get(), &format, _PROFILE_,
+                        icc_profile.data(), icc_profile.size())
+#else
+                    JxlDecoderGetColorAsICCProfile(
+                        dec.get(), _PROFILE_,
+                        icc_profile.data(), icc_profile.size())
+#endif
+                ) {
+                    g_printerr(
+                        "Warning: JxlDecoderGetColorAsICCProfile failed\n");
+                } else {
+                    embProfile = cmsOpenProfileFromMem(icc_profile.data(),
+                                                       icc_profile.size());
+                }
+            } else {
+                g_printerr("Warning: Empty ICC data.\n");
+            }
+        } else if (status == JXL_DEC_NEED_IMAGE_OUT_BUFFER) {
+            format.data_type = JXL_TYPE_FLOAT;
+
+            if (JXL_DEC_SUCCESS !=
+                    JxlDecoderImageOutBufferSize(dec.get(), &format, &buffer_size)) {
+                g_printerr("Error: JxlDecoderImageOutBufferSize failed\n");
+                return IMIO_READERROR;
+            }
+
+            buffer = g_malloc(buffer_size);
+            if (JXL_DEC_SUCCESS != JxlDecoderSetImageOutBuffer(dec.get(), &format, buffer, buffer_size)) {
+                g_printerr("Error: JxlDecoderSetImageOutBuffer failed\n");
+                g_free(buffer);
+                return IMIO_READERROR;
+            }
+        } else if (status == JXL_DEC_FULL_IMAGE ||
+                   status == JXL_DEC_FRAME) {
+            // Nothing to do. If the image is an animation, more full frames
+            // may be decoded. This example only keeps the first one.
+        } else if (status == JXL_DEC_SUCCESS) {
+            // All decoding successfully finished.
+            // It's not required to call JxlDecoderReleaseInput(dec.get())
+            // since the decoder will be destroyed.
+            break;
+        } else if (status == JXL_DEC_NEED_MORE_INPUT) {
+            g_printerr("Error: Already provided all input\n");
```

Comment:
What about the more common "Premature end of file"?

Reply:
The decode loop is from libjxl examples.  Those are the status messages they used. In `jxl/decode.h`, I see nothing corresponding to "premature end of file".  The closest looks like `JXL_DEC_NEED_MORE_INPUT`, but that only occurs when an embedded preview is requested.  Since this isn't asking for a preview, that status would/should never occur.  I guess the result would be a partially decoded image or `JXL_DEC_ERROR`.

Reply:
Okay. If I read "Error: Already provided all input" somewhere this would leave me at a loss. Glad to hear that this should never occur. :sweat_smile:

Reply:
Assuming I'm reading the documentation correctly.  I'll try to test with some truncated files to see what happens.

Reply:
Maybe, `JXL_DEC_NEED_MORE_INPUT` == "premature end of file"